File: tests/lua/nested_match_basic.ml
let a :: b = Cons (a, b)

let rec zip f xs ys =
  match xs, ys with
  | Nil, _ -> [1]
  | _, Nil -> [2]
  | Cons (x, xs), Cons (y, ys) -> f x y :: zip f xs ys

external val ignore : 'a -> () = "nil"
let () = ignore zip

File: tests/lua/nested_match_basic.lua
do
  local Nil = { __tag = "Nil" }
  local function zip(f)
    local function zip_sat(xs, ys)
      if xs.__tag == "Nil" then return { { _1 = 1, _2 = Nil }, __tag = "Cons" } end
      local tmp = xs[1]
      if ys.__tag == "Nil" then return { { _1 = 2, _2 = Nil }, __tag = "Cons" } end
      local tmp0 = ys[1]
      return { { _1 = f(tmp._1)(tmp0._1), _2 = zip_sat(tmp._2, tmp0._2) }, __tag = "Cons" }
    end
    return function(xs) return function(ys) return zip_sat(xs, ys) end end
  end
  (nil)(zip)
end
